// ngx-popups/components/popupsManager.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ngx_dom_component_1 = require("ngx-dom-component");
var popup_component_1 = require("./popup.component");
var ngx_popup_service_1 = require("../services/ngx-popup.service");
// <ngx-popup
// *ngFor="let popupId of popupIds;"
// (popupReady)="onPopupReady($event)"
// (popupClosed)="onPopupClosed($event)"
//   [popupId]="popupId"
//   ></ngx-popup>
var NgxPopupsManagerComponent = /** @class */ (function () {
    function NgxPopupsManagerComponent(popupService, ngxDOMComponentService, element) {
        this.popupService = popupService;
        this.ngxDOMComponentService = ngxDOMComponentService;
        this._popups = [];
        this._element = element.nativeElement;
        if (!this._element.id) {
            this._element.id = 'popups-manager-' + Math.floor(Math.random() * 1e10).toString();
        }
        this.popupService.registerManager(this._element.id, this);
        this.checkVisibility();
    }
    NgxPopupsManagerComponent.prototype.ngAfterViewInit = function () {
        this._ngxDOMComponentContainer = this.ngxDOMComponentService.createContainer(this.popupsContainer);
    };
    NgxPopupsManagerComponent.prototype.open = function (config, waitTransitionEnd, detail) {
        var _this = this;
        return new Promise(function (resolve) {
            var ngxDOMComponent = _this._ngxDOMComponentContainer.create({
                componentType: popup_component_1.NgxPopupComponent
            });
            _this._popups.push(ngxDOMComponent);
            _this.checkVisibility();
            var popup = ngxDOMComponent.instance;
            popup.addEventListener('ready', function () {
                resolve(popup.open(config, waitTransitionEnd, detail).then(function () {
                    return popup;
                }));
            }, { once: true });
            popup.addEventListener('close', function () {
                var index = _this._popups.indexOf(ngxDOMComponent);
                if (index >= 0) {
                    _this._popups[index].destroy();
                    _this._popups.splice(index, 1);
                    _this.checkVisibility();
                }
            }, { once: true });
        });
    };
    NgxPopupsManagerComponent.prototype.close = function (popup, waitTransitionEnd, detail) {
        return popup.close(waitTransitionEnd, detail);
    };
    NgxPopupsManagerComponent.prototype.closeAll = function () {
        return Promise.all(this._popups.slice(0) // clone to avoid removing popups before finishing
            .map(function (popup) {
            return popup.instance.close();
        })).then(function () { return void 0; });
    };
    NgxPopupsManagerComponent.prototype.checkVisibility = function () {
        this._element.classList.toggle('visible', this._popups.length > 0);
    };
    NgxPopupsManagerComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'ngx-popups',
                    template: "\n   <ng-template #popupsContainer></ng-template>\n  "
                },] },
    ];
    /** @nocollapse */
    NgxPopupsManagerComponent.ctorParameters = function () { return [
        { type: ngx_popup_service_1.NgxPopupService, },
        { type: ngx_dom_component_1.NgxDOMComponentService, },
        { type: core_1.ElementRef, },
    ]; };
    NgxPopupsManagerComponent.propDecorators = {
        "popupsContainer": [{ type: core_1.ViewChild, args: ['popupsContainer', { read: core_1.ViewContainerRef },] },],
    };
    return NgxPopupsManagerComponent;
}());
exports.NgxPopupsManagerComponent = NgxPopupsManagerComponent;
