// ngx-popups/ngx-popups.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ngx_dom_component_1 = require("ngx-dom-component");
var popup_component_1 = require("./components/popup.component");
var popupsManager_component_1 = require("./components/popupsManager.component");
var ngx_popup_service_1 = require("./services/ngx-popup.service");
var NgxPopupModule = /** @class */ (function () {
    function NgxPopupModule() {
    }
    NgxPopupModule.decorators = [
        { type: core_1.NgModule, args: [{
                    imports: [ngx_dom_component_1.NgxDOMComponentModule],
                    declarations: [
                        popup_component_1.NgxPopupComponent, popupsManager_component_1.NgxPopupsManagerComponent
                    ],
                    providers: [
                        ngx_popup_service_1.NgxPopupService
                    ],
                    exports: [
                        popup_component_1.NgxPopupComponent, popupsManager_component_1.NgxPopupsManagerComponent
                    ],
                    entryComponents: [
                        popup_component_1.NgxPopupComponent
                    ]
                },] },
    ];
    /** @nocollapse */
    NgxPopupModule.ctorParameters = function () { return []; };
    return NgxPopupModule;
}());
exports.NgxPopupModule = NgxPopupModule;
