// ngx-modal-dialog/src/simple-modal.component.js
import { Component } from '@angular/core';
var SimpleModalComponent = (function () {
    function SimpleModalComponent() {
    }
    /**
     * Initialize dialog with simple HTML content
     * @param {ComponentRef<IModalDialog>} reference
     * @param {Partial<IModalDialogOptions>} options
     */
    SimpleModalComponent.prototype.dialogInit = function (reference, options) {
        if (!options.data) {
            throw new Error("Data information for simple modal dialog is missing");
        }
        this.text = options.data.text;
    };
    return SimpleModalComponent;
}());
export { SimpleModalComponent };
SimpleModalComponent.decorators = [
    { type: Component, args: [{
                selector: 'simple-modal-dialog',
                template: "",
                styles: [':host { display: block; }'],
                host: {
                    '[innerHTML]': 'text'
                }
            },] },
];
/** @nocollapse */
SimpleModalComponent.ctorParameters = function () { return []; };
