// ngx-popups/services/ngx-popup.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var NgxPopupService = /** @class */ (function () {
    function NgxPopupService() {
        this.managers = new Map();
    }
    NgxPopupService.prototype.registerManager = function (id, manager) {
        if (this.managers.get(id)) {
            throw new Error("Duplicate manager id: " + id);
        }
        this.managers.set(id, manager);
    };
    /**
     * Opens a new popup and inject a component inside.
     *
     * @param config - config of the component to inject
     * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
     * @param detail - provide data to the 'detail' property of the custom event 'open'
     * @param managerId - in case of many managers, you can provide a specific manager id
     * @returns {Promise<NgxPopupComponent>} - promise resolved when the popup is opened
     */
    /**
       * Opens a new popup and inject a component inside.
       *
       * @param config - config of the component to inject
       * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
       * @param detail - provide data to the 'detail' property of the custom event 'open'
       * @param managerId - in case of many managers, you can provide a specific manager id
       * @returns {Promise<NgxPopupComponent>} - promise resolved when the popup is opened
       */
    NgxPopupService.prototype.open = /**
       * Opens a new popup and inject a component inside.
       *
       * @param config - config of the component to inject
       * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
       * @param detail - provide data to the 'detail' property of the custom event 'open'
       * @param managerId - in case of many managers, you can provide a specific manager id
       * @returns {Promise<NgxPopupComponent>} - promise resolved when the popup is opened
       */
    function (config, waitTransitionEnd, detail, managerId) {
        if (waitTransitionEnd === void 0) { waitTransitionEnd = true; }
        return this.getManager(managerId).open(config, waitTransitionEnd, detail);
    };
    /**
     * Closes a popup.
     *
     * @param popup - the popup to close
     * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
     * @param detail - provide data to the 'detail' property of the custom event 'open'
     * @returns {Promise<void>} - promise resolved when the popup is closed
     */
    /**
       * Closes a popup.
       *
       * @param popup - the popup to close
       * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
       * @param detail - provide data to the 'detail' property of the custom event 'open'
       * @returns {Promise<void>} - promise resolved when the popup is closed
       */
    NgxPopupService.prototype.close = /**
       * Closes a popup.
       *
       * @param popup - the popup to close
       * @param waitTransitionEnd - (default true) if true wait the end of the animation before triggering open/resolving promise.
       * @param detail - provide data to the 'detail' property of the custom event 'open'
       * @returns {Promise<void>} - promise resolved when the popup is closed
       */
    function (popup, waitTransitionEnd, detail) {
        if (waitTransitionEnd === void 0) { waitTransitionEnd = true; }
        return popup.close(waitTransitionEnd, detail);
    };
    /**
     * Closes all popups.
     *
     * @param managerId
     * @returns {Promise<void>} - promise resolved when all popups are closed.
     */
    /**
       * Closes all popups.
       *
       * @param managerId
       * @returns {Promise<void>} - promise resolved when all popups are closed.
       */
    NgxPopupService.prototype.closeAll = /**
       * Closes all popups.
       *
       * @param managerId
       * @returns {Promise<void>} - promise resolved when all popups are closed.
       */
    function (managerId) {
        return this.getManager(managerId).closeAll();
    };
    NgxPopupService.prototype.getManager = function (id) {
        var manager;
        if (id === void 0) {
            if (this.managers.size > 0) {
                manager = this.managers.values().next().value;
            }
            else {
                throw new Error("No manager for PopupService");
            }
        }
        else if (typeof id === 'string') {
            manager = this.managers.get(id);
            if (manager === void 0) {
                throw new Error("Invalid manager id: " + id);
            }
        }
        else {
            throw new TypeError("Expected string or undefined as id");
        }
        return manager;
    };
    NgxPopupService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    NgxPopupService.ctorParameters = function () { return []; };
    return NgxPopupService;
}());
exports.NgxPopupService = NgxPopupService;
