// ngx-dom-component/ngx-dom-component.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var NgxDOMComponentContainer_1 = require("./classes/NgxDOMComponentContainer");
var NgxDOMComponentComponent = /** @class */ (function () {
    function NgxDOMComponentComponent() {
        this.ngxDOMComponent = null;
        this.ngxDOMComponentContainer = new NgxDOMComponentContainer_1.NgxDOMComponentContainer(this.contentContainerRef);
    }
    NgxDOMComponentComponent.prototype.ngOnChanges = function () {
        if (this.config) {
            if (this.ngxDOMComponent !== null) {
                this.ngxDOMComponent.destroy();
            }
            this.ngxDOMComponent = this.ngxDOMComponentContainer.create(this.config);
        }
    };
    NgxDOMComponentComponent.prototype.ngOnDestroy = function () {
        if (this.ngxDOMComponent !== null) {
            this.ngxDOMComponent.destroy();
            this.ngxDOMComponent = null;
        }
    };
    NgxDOMComponentComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'ngx-dom-component',
                    template: "<ng-template #contentContainer></ng-template>"
                },] },
    ];
    /** @nocollapse */
    NgxDOMComponentComponent.ctorParameters = function () { return []; };
    NgxDOMComponentComponent.propDecorators = {
        "config": [{ type: core_1.Input },],
        "contentContainerRef": [{ type: core_1.ViewChild, args: ['contentContainer', { read: core_1.ViewContainerRef },] },],
    };
    return NgxDOMComponentComponent;
}());
exports.NgxDOMComponentComponent = NgxDOMComponentComponent;
