// ngx-dom-component/classes/MetadataEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONST_PROP_METADATA = '__prop__metadata__';
function SetNgxEntryComponentMetaData(target, metadataEntries) {
    if (!(exports.CONST_PROP_METADATA in target)) {
        Object.defineProperty(target, exports.CONST_PROP_METADATA, {
            value: {},
        });
    }
    var metadataList = target[exports.CONST_PROP_METADATA];
    var metadataEntry;
    for (var i = 0, l = metadataEntries.length; i < l; i++) {
        metadataEntry = metadataEntries[i];
        if (!Array.isArray(metadataList[metadataEntry.propertyName])) {
            metadataList[metadataEntry.propertyName] = [];
        }
        var metadataPropertyList = metadataList[metadataEntry.propertyName];
        // avoid duplicated metaData
        var j = 0, s = metadataPropertyList.length;
        for (; j < s; j++) {
            if (metadataPropertyList[j].ngMetadataName === metadataEntry.metadataName) {
                break;
            }
        }
        if (j === s) {
            metadataPropertyList.push({
                ngMetadataName: metadataEntry.metadataName
            });
        }
    }
}
exports.SetNgxEntryComponentMetaData = SetNgxEntryComponentMetaData;
function GetNgxEntryComponentMetaData(target) {
    var metadataEntries = [];
    var metadataList;
    if ((typeof Reflect !== 'undefined') && (typeof Reflect.getOwnMetadata === 'function')) {
        metadataList = Reflect.getOwnMetadata('propMetadata', target);
        if (metadataList) {
            var _loop_1 = function (propertyName) {
                if (metadataList[propertyName].length > 0) {
                    metadataList[propertyName].forEach(function (metadata) {
                        var metadataName = metadata.toString();
                        if (metadataName.startsWith('@')) {
                            metadataName = metadataName.substring(1);
                        }
                        metadataEntries.push({
                            propertyName: propertyName,
                            metadataName: metadataName
                        });
                    });
                }
            };
            for (var propertyName in metadataList) {
                _loop_1(propertyName);
            }
        }
    }
    metadataList = target['__prop__metadata__'];
    if (metadataList) {
        var _loop_2 = function (propertyName) {
            if (metadataList[propertyName].length > 0) {
                metadataList[propertyName].forEach(function (metadata) {
                    metadataEntries.push({
                        propertyName: propertyName,
                        metadataName: metadata.ngMetadataName
                    });
                });
            }
        };
        for (var propertyName in metadataList) {
            _loop_2(propertyName);
        }
    }
    return metadataEntries;
}
exports.GetNgxEntryComponentMetaData = GetNgxEntryComponentMetaData;
